src/main/java/org/truffleruby/core/basicobject/BasicObjectNodes.java

```diff
@@ -126,11 +126,16 @@ public boolean equal(DynamicObject a, DynamicObject b) {
             return a == b;
         }
 
-        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "notSameClass(a, b)", "isNotIntLong(a) || isNotIntLong(b)" })
+        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "!sameClass(a, b)", "isNotIntLong(a) || isNotIntLong(b)" })
         public boolean equalIncompatiblePrimitiveTypes(Object a, Object b) {
             return false;
         }
 
+        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "sameClass(a, b)", "isNotIntLongDouble(a) || isNotIntLongDouble(b)" })
```

Comment:
If they are the same class, you only need to test one for isNotIntLongDouble().
But this test does not seem correct, because we could have byte/char/short coming in and they would wrongly compare by identity. `!isBoxedPrimitive(a)` should be used instead.
Another and imho better way to fix this would be to properly convert incoming primitives smaller than int to int when they come in Ruby.